util/workloadidentity/workloadidentity.go

```diff
@@ -29,17 +35,29 @@ func NewWorkloadIdentityTokenProvider() TokenProvider {
 	return WorkloadIdentityTokenProvider{tokenCredential: cred}
 }
 
-func (c WorkloadIdentityTokenProvider) GetToken(scope string) (string, error) {
+func (c WorkloadIdentityTokenProvider) GetToken(scope string) (*Token, error) {
 	if initError != nil {
-		return "", initError
+		return nil, initError
 	}
 
 	token, err := c.tokenCredential.GetToken(context.Background(), policy.TokenRequestOptions{
 		Scopes: []string{scope},
 	})
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	return token.Token, nil
+	return &Token{AccessToken: token.Token, ExpiresOn: token.ExpiresOn}, nil
+}
+
+func CalclulateCacheExpiryBasedOnTokenExpiry(tokenExpiry time.Time) time.Duration {
+	// Calculate the cache expiry as 5 minutes before the token expires
+	cacheExpiry := time.Until(tokenExpiry) - time.Minute*5
+
+	// If the time until the token expiry (minus 5 minutes)
+	// is less than or equal to zero, set the expiry to one second.
```

Comment:
Why one second?

Reply:
THis expiry time is set in the gocache. Incase it is expiry is negative the item will never expire. If the expiry is 0 then it will use the default expiry which is never.

Hence to avoid this it is minimum set to 1 second to allow the token cache expiry.

Reply:
Gotcha. Instead of returning a very-small expiry, I think we should just return 0 and have the caller skip setting the cache value. We shouldn't add to the cache when we have no intention to use the cached value.

Reply:
done